src/serialization.rs

```diff
@@ -0,0 +1,171 @@
+pub const MAX_VEC_SIZE: usize = 4_000_000;
+
+#[derive(Debug)]
+pub enum SerialisationError {
+    IO(std::io::Error),
+
+    ConsensusEcode(bitcoin::consensus::encode::Error),
+
+    Secp256k1(bitcoin::secp256k1::Error),
+
+    Key(bitcoin::util::key::Error),
+
+    OverSizedAllocation { requested: usize, max: usize },
+
+    General(&'static str),
+}
+
+impl From<bitcoin::secp256k1::Error> for SerialisationError {
+    fn from(e: bitcoin::secp256k1::Error) -> Self {
+        SerialisationError::Secp256k1(e)
+    }
+}
+
+impl From<bitcoin::consensus::encode::Error> for SerialisationError {
+    fn from(e: bitcoin::consensus::encode::Error) -> Self {
+        SerialisationError::ConsensusEcode(e)
+    }
+}
+
+impl From<std::io::Error> for SerialisationError {
+    fn from(e: std::io::Error) -> Self {
+        SerialisationError::IO(e)
+    }
+}
+
+impl From<bitcoin::util::key::Error> for SerialisationError {
+    fn from(e: bitcoin::util::key::Error) -> Self {
+        SerialisationError::Key(e)
+    }
+}
+
+// A Serializable Trait that defines byte encoding of structures
+// Anywhere we need byte encoded data (Network, Database) we can use this.
+// The underlying encoding of primitve datatyes and structures are done
+// via bitcoin::consensus::encode::Encodale trait.
+pub trait Serializable {
+    fn serialise<W: std::io::Write>(&self, w: W) -> Result<usize, SerialisationError>;
+}
+
+pub trait Deserilizable: Sized {
+    fn deserialise<R: std::io::Read>(r: R) -> Result<Self, SerialisationError>;
+}
+
+// Implementation of Ser/Deserializable for some primitives not covered
+// by bitcoin::consensus::encode::Encodable
+impl Serializable for [u8; 20] {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        w.emit_slice(&self[..])?;
+        Ok(20)
+    }
+}
+
+impl Deserilizable for [u8; 20] {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut bytes = [0u8; 20];
+        r.read_slice(&mut bytes)?;
+
+        Ok(bytes)
+    }
+}
+
+impl Serializable for Signature {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(w.write(&self.serialize_compact()[..])?)
+    }
+}
+
+impl Deserilizable for Signature {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut sig_buff = [0u8; 64];
+        r.read_exact(&mut sig_buff)?;
+        Ok(Signature::from_compact(&sig_buff[..])?)
+    }
+}
+
+impl Serializable for SecretKey {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(w.write(&self[..])?)
+    }
+}
+
+impl Deserilizable for SecretKey {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut secret_key_bytes = [0u8; 32];
+        r.read_slice(&mut secret_key_bytes)?;
+        Ok(SecretKey::from_slice(&secret_key_bytes)?)
+    }
+}
```

Comment:
[rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/key.rs#L218-L242) has a `serde` implementation for `SecretKey`